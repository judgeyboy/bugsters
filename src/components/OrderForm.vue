<template>
  <div>

    <p>
      Use the slider to select how much you would like to invest in testing your application
      and provide us with your details. We'll do the rest. Of course we'll send you an offer...
    </p>

    <form class="form-horizontal"
          @submit.prevent="handleSubmit('order-form')"
          data-vv-scope="order-form">

      <div class="form-group">
        <div class="col-3 col-xs-12">
          <label class="form-label" for="order-budget">Budget</label>
        </div>
        <div class="col-6 col-xs-8">
          <input class="slider"
                 type="range"
                 max="2500"
                 min="200"
                 step="100"
                 v-model="budget" />
        </div>
        <div class="col-3 col-xs-4">
          <label class="form-label text-center"><strong>{{ budget }} .- CHF</strong></label>
        </div>
      </div>

      <div class="form-group">
        <div class="col-3 col-xs-12">
          <label class="form-label" for="order-company">Company</label>
        </div>
        <div class="col-9 col-xs-12">
          <input type="text"
                 id="order-company"
                 class="form-input"
                 name="company"
                 placeholder="company"
                 v-model="company"
                 v-validate="'required'"
                 :class="{'is-error': errors.has('order-form.company')}">
        </div>
      </div>

      <div class="form-group">
        <div class="col-3 col-xs-12">
          <label class="form-label" for="order-name">Name</label>
        </div>
        <div class="col-9 col-xs-12">
          <input type="text"
                 id="order-name"
                 class="form-input"
                 name="name"
                 placeholder="name"
                 v-model="name"
                 v-validate="'required'"
                 :class="{'is-error': errors.has('order-form.name')}">
        </div>
      </div>

      <div class="form-group">
        <div class="col-3 col-xs-12">
          <label class="form-label" for="order-email">Email</label>
        </div>
        <div class="col-9 col-xs-12">
          <input type="email"
                 id="order-email"
                 class="form-input"
                 name="email"
                 placeholder="email"
                 v-model="email"
                 v-validate="'required|email'"
                 :class="{'is-error': errors.has('order-form.email')}">
        </div>
      </div>

      <div class="form-group">
        <div class="col-3 col-xs-12">
          <label class="form-label" for="order-phone">Phone</label>
        </div>
        <div class="col-9 col-xs-12">
          <input type="text"
                 id="order-phone"
                 class="form-input"
                 name="phone"
                 placeholder="phone"
                 v-model="phone"
                 v-validate="'required'"
                 :class="{'is-error': errors.has('order-form.phone')}">
        </div>
      </div>

      <div class="form-group">
        <div class="col-3 col-xs-12">
          <label class="form-label" for="order-url">Page Url</label>
        </div>
        <div class="col-9 col-xs-12">
          <input type="text"
                 id="order-url"
                 class="form-input"
                 name="url"
                 placeholder="url"
                 v-model="url"
                 v-validate="'required|url'"
                 :class="{'is-error': errors.has('order-form.url')}">
        </div>
      </div>

      <div class="form-group">
        <div class="col-3 col-xs-12">
          <label class="form-label" for="order-message">Message</label>
        </div>
        <div class="col-9 col-xs-12">
          <textarea placeholder="message"
                    id="order-message"
                    class="form-input"
                    name="message"
                    v-model="message"
                    v-validate="'max:200'"
                    :class="{'is-error': errors.has('order-form.message')}"></textarea>
        </div>
      </div>
      <div class="form-buttons">
        <button type="submit" value="Send" class="btn btn-primary" :class="{loading: isLoading}">Send</button>
      </div>
    </form>
  </div>
</template>

<script>

export default {
  name: 'OrderForm',

  data () {
    return {
      budget: 500,
      company: '',
      name: '',
      email: '',
      url: '',
      message: '',
      phone: '',
      isLoading: false
    }
  },

  inject: ['$validator'],

  methods: {
    handleSubmit (scope) {
      return this.$validator.validateAll(scope).then(valid => {
        if (!valid) {
          alert('fix dem errors')
          return
        }

        this.isLoading = true

        const orderDetails = {
          budget: this.budget,
          company: this.company,
          name: this.name,
          email: this.email,
          url: this.url,
          message: this.message,
          phone: this.phone
        }

        this.$emit('submit', orderDetails)
      })
    }
  }
}
</script>
