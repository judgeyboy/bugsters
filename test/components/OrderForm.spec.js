import VeeValidate, { Validator } from 'vee-validate'
import { shallow, createLocalVue } from 'vue-test-utils'

import OrderForm from '@/components/OrderForm'

const localVue = createLocalVue()

localVue.use(VeeValidate)

describe('OrderForm component', () => {
  it('should be a vue instance', () => {
    const wrapper = shallow(OrderForm, {
      localVue,
      provide: {
        $validator: new Validator()
      }
    })

    expect(wrapper.isVueInstance()).toBe(true)
  })

  it('should call handleSubmit when clicking the send button', () => {
    const wrapper = shallow(OrderForm, {
      localVue,
      provide: {
        $validator: new Validator()
      }
    })

    let spy = jest.spyOn(wrapper.vm, 'handleSubmit')

    const submitButton = wrapper.find('button[type=submit]')
    submitButton.trigger('click')

    expect(spy).toHaveBeenCalledWith('order-form')
  })

  it('should emit "submit" event with order details', async () => {
    const wrapper = shallow(OrderForm, {
      localVue,
      provide: {
        $validator: new Validator()
      }
    })
    let spy = jest.spyOn(wrapper.vm, '$emit')

    const orderDetails = {
      budget: 500,
      company: 'International Spying Agency',
      name: 'Sterling Archer',
      email: 'duchess@spy-agency.com',
      url: 'https://sterling-malory-archer.com',
      message: 'This is an awesome message',
      phone: '0041794672849'
    }

    wrapper.setData(orderDetails)

    await wrapper.vm.handleSubmit('order-form')

    expect(spy).toHaveBeenCalledWith('submit', orderDetails)
  })

  it('should not emit "submit" event if the form has errors', async () => {
    const wrapper = shallow(OrderForm, {
      localVue,
      provide: {
        $validator: new Validator()
      }
    })

    let spy = jest.spyOn(wrapper.vm, '$emit')

    const invalidOrderDetails = {
      budget: 1200,
      company: 'International Spying Agency',
      name: 'Sterling Archer',
      email: 'duchess@spy-agency.com',
      url: '',
      message: 'This is an awesome message'
    }

    wrapper.setData(invalidOrderDetails)

    await wrapper.vm.handleSubmit('order-form')

    expect(spy).not.toHaveBeenCalled()
  })
})
