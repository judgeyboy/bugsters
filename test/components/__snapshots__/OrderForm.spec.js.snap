// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderForm component snapshot 1`] = `

<div>
  <p>
    Use the slider to select how much you would like to invest in testing your application
    and provide us with your details. We'll do the rest. Of course we'll send you an offer...
  </p>
  <form data-vv-scope="order-form"
        class="form-horizontal"
  >
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-budget"
               class="form-label"
        >
          Budget
        </label>
      </div>
      <div class="col-6 col-xs-8">
        <input type="range"
               max="2500"
               min="200"
               step="100"
               class="slider"
        >
      </div>
      <div class="col-3 col-xs-4">
        <label class="form-label text-center">
          <strong>
            500 .- CHF
          </strong>
        </label>
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-company"
               class="form-label"
        >
          Company
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-company"
               name="company"
               placeholder="company"
               class="form-input"
               data-vv-id="31"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-name"
               class="form-label"
        >
          Name
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-name"
               name="name"
               placeholder="name"
               class="form-input"
               data-vv-id="32"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-email"
               class="form-label"
        >
          Email
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="email"
               id="order-email"
               name="email"
               placeholder="email"
               class="form-input"
               data-vv-id="33"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-phone"
               class="form-label"
        >
          Phone
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-phone"
               name="phone"
               placeholder="phone"
               class="form-input"
               data-vv-id="34"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-url"
               class="form-label"
        >
          Page Url
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-url"
               name="url"
               placeholder="url"
               class="form-input"
               data-vv-id="35"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-message"
               class="form-label"
        >
          Message
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <textarea placeholder="message"
                  id="order-message"
                  name="message"
                  class="form-input"
                  data-vv-id="36"
                  aria-required="false"
                  aria-invalid="false"
        >
        </textarea>
      </div>
    </div>
    <div class="form-buttons">
      <button type="submit"
              value="Send"
              class="btn btn-primary"
      >
        Send
      </button>
    </div>
  </form>
</div>

`;

exports[`OrderForm component snapshot with cancel button 1`] = `

<div>
  <p>
    Use the slider to select how much you would like to invest in testing your application
    and provide us with your details. We'll do the rest. Of course we'll send you an offer...
  </p>
  <form data-vv-scope="order-form"
        class="form-horizontal"
  >
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-budget"
               class="form-label"
        >
          Budget
        </label>
      </div>
      <div class="col-6 col-xs-8">
        <input type="range"
               max="2500"
               min="200"
               step="100"
               class="slider"
        >
      </div>
      <div class="col-3 col-xs-4">
        <label class="form-label text-center">
          <strong>
            500 .- CHF
          </strong>
        </label>
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-company"
               class="form-label"
        >
          Company
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-company"
               name="company"
               placeholder="company"
               class="form-input"
               data-vv-id="43"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-name"
               class="form-label"
        >
          Name
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-name"
               name="name"
               placeholder="name"
               class="form-input"
               data-vv-id="44"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-email"
               class="form-label"
        >
          Email
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="email"
               id="order-email"
               name="email"
               placeholder="email"
               class="form-input"
               data-vv-id="45"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-phone"
               class="form-label"
        >
          Phone
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-phone"
               name="phone"
               placeholder="phone"
               class="form-input"
               data-vv-id="46"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-url"
               class="form-label"
        >
          Page Url
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-url"
               name="url"
               placeholder="url"
               class="form-input"
               data-vv-id="47"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-message"
               class="form-label"
        >
          Message
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <textarea placeholder="message"
                  id="order-message"
                  name="message"
                  class="form-input"
                  data-vv-id="48"
                  aria-required="false"
                  aria-invalid="false"
        >
        </textarea>
      </div>
    </div>
    <div class="form-buttons">
      <a class="btn cancel-button">
        Cancel
      </a>
      <button type="submit"
              value="Send"
              class="btn btn-primary"
      >
        Send
      </button>
    </div>
  </form>
</div>

`;

exports[`OrderForm component snapshot with form error 1`] = `

<div>
  <p>
    Use the slider to select how much you would like to invest in testing your application
    and provide us with your details. We'll do the rest. Of course we'll send you an offer...
  </p>
  <form data-vv-scope="order-form"
        class="form-horizontal"
  >
    <div>
      <span>
        Form contains errors, please check the red fields.
      </span>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-budget"
               class="form-label"
        >
          Budget
        </label>
      </div>
      <div class="col-6 col-xs-8">
        <input type="range"
               max="2500"
               min="200"
               step="100"
               class="slider"
        >
      </div>
      <div class="col-3 col-xs-4">
        <label class="form-label text-center">
          <strong>
            500 .- CHF
          </strong>
        </label>
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-company"
               class="form-label"
        >
          Company
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-company"
               name="company"
               placeholder="company"
               class="form-input"
               data-vv-id="37"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-name"
               class="form-label"
        >
          Name
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-name"
               name="name"
               placeholder="name"
               class="form-input"
               data-vv-id="38"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-email"
               class="form-label"
        >
          Email
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="email"
               id="order-email"
               name="email"
               placeholder="email"
               class="form-input"
               data-vv-id="39"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-phone"
               class="form-label"
        >
          Phone
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-phone"
               name="phone"
               placeholder="phone"
               class="form-input"
               data-vv-id="40"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-url"
               class="form-label"
        >
          Page Url
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <input type="text"
               id="order-url"
               name="url"
               placeholder="url"
               class="form-input"
               data-vv-id="41"
               aria-required="true"
               aria-invalid="false"
        >
      </div>
    </div>
    <div class="form-group">
      <div class="col-3 col-xs-12">
        <label for="order-message"
               class="form-label"
        >
          Message
        </label>
      </div>
      <div class="col-9 col-xs-12">
        <textarea placeholder="message"
                  id="order-message"
                  name="message"
                  class="form-input"
                  data-vv-id="42"
                  aria-required="false"
                  aria-invalid="false"
        >
        </textarea>
      </div>
    </div>
    <div class="form-buttons">
      <button type="submit"
              value="Send"
              class="btn btn-primary"
      >
        Send
      </button>
    </div>
  </form>
</div>

`;
